File: apps/web/src/config/cloudFunctions.ts
/**
 * CloudFunctions — the deployed callable names and the region they live in.
 *
 * WHY names and not URLs: these are `onCall` callables, so the client references them by NAME and the
 * callable protocol resolves the URL. What it CANNOT resolve is the region.
 *
 * WHY the region is the important line in this file: `functions/src/index.ts` sets
 * `setGlobalOptions({ region: 'asia-south1' })` so the onDocument* triggers sit in the same region as
 * the Firestore database. A `getFunctions(app)` with no region silently targets `us-central1`, where
 * nothing is deployed — every call 404s and no function log is written, because the request never
 * reaches one. Mirrors `apps/app/lib/core/constants/cloud_functions.dart`.
 *
 * These names are verified against the deployed functions, not against the spec.
 */
export const CloudFunctions = {
  region: 'asia-south1',

  // --- Auth (MSG91 OTP — pre-login, deliberately NO auth gate) ---
  //
  // THE WEBSITE MUST SEND `source: 'web'` ON authOTP. The site and the app use different MSG91
  // widgets; the server picks the widget from this field and remembers it, so verifyOTP/resendOTP hit
  // the SAME widget the code was issued by (a reqId is not portable across widgets). Omitting it
  // resolves to the APP widget — a code the customer never received on the web widget, so verify then
  // fails for no visible reason. verifyOTP/resendOTP take no source: the server reads it back from the
  // pending request. This is the one field W3 cannot leave off.
  OTP_SOURCE: 'web',
  /** Sends the OTP. `{ phone: '10 digits', countryCode: '+91', source: 'web' }` -> `{ success, message }`. */
  authOTP: 'authOTP',
  /**
   * `{ phone, countryCode, otp }` -> `{ verified: true, customToken, isNewUser }` | `{ verified: false }`.
   * `verified: false` is a WRONG CODE, not an error — show it inline, do not throw.
   */
  verifyOTP: 'verifyOTP',
  /** `{ phone, countryCode, retryType: 'text' | 'voice' }`. */
  resendOTP: 'resendOTP',

  // --- Orders & payment ---
  createPaymentOrder: 'createPaymentOrder',
  verifyPayment: 'verifyPayment',
  generateInvoicePDF: 'generateInvoicePDF',
  cancelOrder: 'cancelOrder',

  // --- Wallet top-up (two steps: the amount is fixed server-side BEFORE payment, and the credit
  //     happens only against a verified signature afterwards). ---
  createWalletTopUpOrder: 'createWalletTopUpOrder',
  verifyWalletTopUp: 'verifyWalletTopUp',

  // --- Growth ---
  spinWheel: 'spinWheel',

  /**
   * ADMIN ONLY (`validateAdmin`). Listed so the name is documented in one place — a customer calling
   * it gets permission-denied. The customer top-up path is the two functions above.
   */
  addWalletMoney: 'addWalletMoney',
} as const;

File: apps/app/lib/core/constants/cloud_functions.dart
/// Deployed Cloud Function identifiers for the customer app.
///
/// NOTE: the skill's `cloud_function_urls.dart` pattern assumes HTTPS-triggered
/// functions with unique Cloud Run URLs. The Barakath functions are **callables**
/// (`onCall`) deployed to `asia-south1`, so the client references them by NAME
/// (the callable protocol resolves the name) and must region-pin the instance —
/// `FirebaseFunctions.instanceFor(region: CloudFunctions.region)`. A default
/// (us-central1) instance would 404. Hence names + a region here, not URLs.
class CloudFunctions {
  const CloudFunctions._();

  static const String region = 'asia-south1';

  // Order & payment
  static const String createPaymentOrder = 'createPaymentOrder';
  static const String verifyPayment = 'verifyPayment';
  static const String generateInvoicePDF = 'generateInvoicePDF';

  // Auth (MSG91 OTP — pre-login, no auth required)
  static const String authOTP = 'authOTP';
  static const String verifyOTP = 'verifyOTP';
  static const String resendOTP = 'resendOTP';

  /// Cancels an order and refunds it (spec §5.2).
  ///
  /// NOTE: spec §5.2 describes this as admin-auth, but §2.17 gives the customer
  /// a "Cancel item" action at Pending. If the callable rejects a customer
  /// caller, widen the check server-side — do NOT cancel from the client.
  static const String cancelOrder = 'cancelOrder';

  // Growth
  static const String spinWheel = 'spinWheel';

  /// ADMIN-ONLY (`validateAdmin`): the admin panel crediting a customer's
  /// wallet. NOT the customer top-up path and not callable from this app — a
  /// customer calling it gets permission-denied. Kept here only so the name is
  /// documented in one place; use the two below instead.
  static const String addWalletMoney = 'addWalletMoney';

  // Customer wallet top-up (spec §2.20 "Add money"). Two steps, because the
  // amount must be fixed server-side BEFORE payment and the credit must happen
  // only against a verified signature afterwards.
  /// Opens a bare Razorpay order and records what it is for.
  static const String createWalletTopUpOrder = 'createWalletTopUpOrder';

  /// Verifies the Razorpay signature and credits the wallet. Idempotent — safe
  /// to retry after a dropped connection.
  static const String verifyWalletTopUp = 'verifyWalletTopUp';
}
